pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go

```diff
@@ -63,19 +58,18 @@ func ProvideDynamic(cfg *config.Cfg, store angularpatternsstore.Service, feature
 		httpClient:            makeHttpClient(),
 		baseURL:               cfg.GrafanaComURL,
 		backgroundJobInterval: defaultBackgroundJobInterval,
-		initialRestore:        make(chan struct{}),
 	}
 
 	// Perform the initial restore from db without blocking
+	d.mux.Lock()
```

Comment:
what if you don't `Lock` here but in `setDetectorsFromCache`? I am thinking on something like:

```patch
func (d *Dynamic) setDetectorsFromCache(ctx context.Context) error {
	var cachedPatterns GCOMPatterns
+ d.mux.RLock()
	rawCached, ok, err := d.store.Get(ctx)
+ d.mux.RUnlock()
	if !ok {
		// No cached value found, do not alter in-memory detectors
		return nil
	}
	if err != nil {
		return fmt.Errorf("get cached value: %w", err)
	}
	// Try to unmarshal, convert to detectors and set local cache
	if err := json.Unmarshal([]byte(rawCached), &cachedPatterns); err != nil {
		return fmt.Errorf("json unmarshal: %w", err)
	}
	cachedDetectors, err := d.patternsToDetectors(cachedPatterns)
	if err != nil {
		return fmt.Errorf("convert to detectors: %w", err)
	}
-	// Do not use setDetectors or it will cause a deadlock
- 	d.detectors = cachedDetectors
+ d.setDetectors()
	return nil
}
```

Reply:
Unfortunately this doesn't work. The first call of ProvideDetectors must await the initial restore, that's why I had the channel in the first place.

If we want to use a mutex, we have to lock it before launching the [goroutine that fetches the values from the database in the background](https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go#L58), otherwise whoever calls ProvideDetectors may find it unlocked (since there's no guarantee the "restore" goroutine will run first) and return an empty value (which is wrong).

I also have [a test](https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/pluginsintegration/angulardetectorsprovider/dynamic_test.go#L67) for this, with your suggested changes the test fails.

Alternatively, we can perform the restore in the main goroutine, but this will increase the Grafana startup times significantly if GCOM is slow to respond:

```diff
--- a/pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go
+++ b/pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go
@@ -52,17 +52,11 @@ func ProvideDynamic(cfg *config.Cfg, store angularpatternsstore.Service, feature
                httpClient: makeHttpClient(),
                baseURL:    cfg.GrafanaComURL,
        }
-
-       // Perform the initial restore from db without blocking
-       d.mux.Lock()
-       go func() {
-               d.log.Debug("Restoring cache")
-               if err := d.setDetectorsFromCache(context.Background()); err != nil {
-                       d.log.Warn("Cache restore failed", "error", err)
-               }
-               d.mux.Unlock()
-       }()
-
+       // Perform the initial restore from db, return the service once restore is done
+       d.log.Debug("Restoring cache")
+       if err := d.setDetectorsFromCache(context.Background()); err != nil {
+               d.log.Warn("Cache restore failed", "error", err)
+       }
        return d, nil
 }
 
@@ -175,8 +169,7 @@ func (d *Dynamic) setDetectorsFromCache(ctx context.Context) error {
        if err != nil {
                return fmt.Errorf("convert to detectors: %w", err)
        }
-       // Do not use setDetectors or it will cause a deadlock
-       d.detectors = cachedDetectors
+       d.setDetectors(cachedDetectors)
        return nil
 }
```

I would either keep it like it is currently (locking before launching the other goroutine), or revert to using a channel if it's more confusing (I personally think using a channel is more idiomatic, but no strong opinions on this)

Reply:
> Alternatively, we can perform the restore in the main goroutine, but this will increase the Grafana startup times significantly if GCOM is slow to respond

I may not be seeing it but there is no GCOM request in `setDetectorsFromCache`? Only getting from the `store` which should be fast?

Reply:
Sorry I meant database not GCOM. If we can assume that the database is always fast, then sure, I will change it

Reply:
:+1: yes, I don't think we need to worry about that